File: subroutine_add_hydrogen.f
function normalize(vec) result(nvec)!{{{
real(kind=rk),intent(in) :: vec(:)
real(kind=rk) :: nvec(1:3)
nvec=vec/sqrt(sum(vec**2))
end function normalize!}}}
    
function v2q(v,theta) result(q)
!Create a quaternion from the vector v and angle theta
real(kind=rk),intent(in) :: v(:),theta ! Rad
real(kind=rk) :: q(4)
q=[cos(theta/2),sin(theta/2)*normalize(v)]
end function v2q

function RV(q) result(m)
!Rotational matrix: B.Stevensson et. al. 2011
real(kind=rk),intent(in) :: q(:) !quat
real(kind=rk) :: m(3,3) !quat
!First col
m(1,1)=q(1)**2+q(2)**2-q(3)**2-q(4)**2
m(2,1)=2*(q(2)*q(3)+q(1)*q(4))
m(3,1)=2*(q(2)*q(4)-q(1)*q(3))
!Second col
m(1,2)=2*(q(2)*q(3)-q(1)*q(4))
m(2,2)=q(1)**2-q(2)**2+q(3)**2-q(4)**2
m(3,2)=2*(q(3)*q(4)+q(1)*q(2))
!Third col
m(1,3)=2*(q(2)*q(4)+q(1)*q(3))
m(2,3)=2*(q(3)*q(4)-q(1)*q(2))
m(3,3)=q(1)**2-q(2)**2-q(3)**2+q(4)**2
end function RV


subroutine add_hydrogen(helpers,hcoor,imol,func,bondlength,theta)
    integer(kind=ik),intent(in) :: helpers(:),func,imol
    integer(kind=ik) :: i
    real(kind=rk) ::&
    v1(3),v2(3),v3(3),v4(3),v5(3),u(3),bondlength,thetal
    real(kind=rk),intent(inout) :: hcoor(:)
    real(kind=rk),optional :: theta
    if(.not.present(theta) .and. func==3)stop 'CH2x needs rotation angle'
    select case(func)
    case(1) ! CH
        v1=atom(helpers(1))%coor(:,imol)
        v2=0._rk
        do i=2,size(helpers)
        ! Center of the nomalized vectors
        v2=v2+normalize(atom(helpers(i))%coor(:,imol)-v1)
        end do
        ! Normalize by number and shift it back to v1
        v2=v2/(size(helpers)-1)+v1
        hcoor=bondlength*normalize(v1-v2)+v1
    case(2) !CH double bond
        v1=atom(helpers(1))%coor(:,imol)
        v2=normalize(atom(helpers(2))%coor(:,imol)-v1)
        v3=normalize(atom(helpers(3))%coor(:,imol)-v1)
        ! thetal is the angle 2pi - C-C-C devided by 2
        ! to ensure equal (C-C-H) angles from both directions
        thetal=pi*(2-valenceangle(helpers(1),&
                helpers(2),&
                helpers(3),imol)/180._rk)/2
        u=normalize(cross_product(v2,v3))
        ! RV returns a rotational 3x3 matrix from a quaternion
        ! v2q turns a vector into a quaternion with angle theta
        ! Rotate by matrix multiplication RV*V3 where V3 is rotated
        hcoor=bondlength*normalize(matmul(RV(v2q(u,thetal)),v3))+v1
    case(3) !CH2
        thetal=theta*pi/180._rk
        v1=atom(helpers(1))%coor(:,imol)
        v2=normalize(atom(helpers(2))%coor(:,imol)-v1)
        v3=normalize(atom(helpers(3))%coor(:,imol)-v1)
        ! Perpendicular to the C-C-C plane
        v4=normalize(cross_product(v3,v2))
        ! Vector to rotate around. In the plane.
        u=normalize(v2-v3) 
        ! The vector to be rotated theta/2, perpendicular to u and v4
        v4=normalize(cross_product(v4,u))
        ! Rotate the new v4 around u by thetal 
        hcoor=bondlength*normalize(matmul(RV(v2q(u,thetal)),v4))+v1
    case(4) !CH3e
        thetal=acos(-1._rk/3) ! 109.47 deg in radians
        v1=atom(helpers(1))%coor(:,imol)
        v2=normalize(atom(helpers(2))%coor(:,imol)-v1)
        v3=normalize(atom(helpers(3))%coor(:,imol)-v1)
        u=normalize(cross_product(v3,v2))
        hcoor=bondlength*normalize(matmul(RV(v2q(u,thetal)),v2))+v1
    case(5) !CH3r
        thetal=120._rk*pi/180._rk
        v1=atom(helpers(1))%coor(:,imol)
        v2=atom(helpers(2))%coor(:,imol)
        v3=atom(helpers(3))%coor(:,imol) !First hydrogen added by CH3e
        u=normalize(v2-v1) ! C-C bond to rotate around
        v4=normalize(v3-v1) 
        hcoor=bondlength*normalize(matmul(RV(v2q(u,thetal)),v4))+v1
    case(6) !CH3s
        thetal=-120._rk*pi/180._rk
        v1=atom(helpers(1))%coor(:,imol)
        v2=atom(helpers(2))%coor(:,imol)
        v3=atom(helpers(3))%coor(:,imol) !First hydrogen added by CH3e
        u=normalize(v2-v1) ! C-C bond to rotate around
        v4=normalize(v3-v1)
        hcoor=bondlength*normalize(matmul(RV(v2q(u,thetal)),v4))+v1
    end select
end subroutine add_hydrogen

File: subroutine_add_hydrogen.py
#!/usr/bin/env python


import numpy as np 

def normalize(vec):
    """Function that normalize a 3-D vector
	Take in a numpy array of dimension 3
	Return a normalized numpy array of dimension 3"""

    nvec = vec / np.sqrt(np.sum(vec**2))
    return nvec

def v2q(vec, theta):
    """Function that translate a 3-D vector and angle theta in a quaternion.
	Takes in a numpy array of dimension 3 and a theta angle (in radian)
	Returns a quaternion; numpy array of dimension 4"""

    w = np.cos(theta/2)
    (x, y, z) = np.sin(theta/2)*normalize(vec)
    q = np.array([w, x, y, z])
    return q


def rotational_matrix(quaternion):
    """Function that translate de quaternion into a rotational matrix
	Takes in a numpy array of dimension 4 i.e a quaternion
	Returns a rotational matrix; numpey array dimension [3,3]"""
    #init mat_rot np.array of dim 3,3 default values set to 0
    mat_rot = np.zeros([3,3])
    w = quaternion[0]
    x = quaternion[1]
    y = quaternion[2]
    z = quaternion[3]
    mat_rot[0,0] = w**2 + x**2 - y**2 - z**2
    mat_rot[1,0] = 2*(x*y + w*z)
    mat_rot[2,0] = 2*(x*z - w*y)
    mat_rot[0,1] = 2*(x*y - w*z)
    mat_rot[1,1] = w**2 - x**2 + y**2 - z**2
    mat_rot[2,1] = 2*(y*z + w*x)
    mat_rot[0,2] = 2*(x*z + w*y)
    mat_rot[1,2] = 2*(y*z - w*x)
    mat_rot[2,2] = w**2 - x**2 - y**2 + z**2
    return mat_rot

#    1DOPC    N4    4   4.389   7.455   2.167
#    1DOPC    C5    5   4.438   7.592   2.185
#    1DOPC   H51    6   4.357   7.651   2.181
#    1DOPC   H52    7   4.497   7.611   2.106
#    1DOPC    C6    8   4.516   7.627   2.312

def apply_rotation(vec_to_rotate, rotational_vector, rad_angle):
    """Function that take a vector to rotate around an other vector 
    by a given angle. It returns the final vector normalised"""
    #generate a quaternion of the given angle (in radian)
    quat_rot = v2q(rotational_vector, rad_angle)
    #generate the rotational matrix 
    rot_mat_quat = rotational_matrix(quat_rot)
    #Use the rotational matrix on the vector to rotate
    vec_rotated = np.dot(rot_mat_quat, vec_to_rotate)
    norm_vec = normalize(vec_rotated)
    return norm_vec
    

def write_PDB(id, atom_type, coor):
    """Function that print information in a PDB format
    Helps check the result with VMD """
    x = coor[0]
    y = coor[1]
    z = coor[2]
    #From cupnet website ! 
    print "{:6s}{:5d} {:^4s}{:1s}{:3s} {:1s}{:4d}{:1s}   {:8.3f}{:8.3f}{:8.3f}{:6.2f}{:6.2f}          {:>2s}{:2s}".format("ATOM",id, atom_type, "", "LIP", "A", 1,"",  x, y, z, 1.0, 0.0, "", "")
    
    
def get_SP2_H(atom, helper1, helper2):
    """Function that reconstruct the 2 hydrogens of a SP2 carbon
    It needs the carbon coordinate and 2 helpers, the previous atom and 
    the next one.
    It returns the coordinates of the hydrogens in a tuple of 3D list 
    ([x_H1, y_H1, z_H1], [x_H2, y_H2, z_H2])
    """
    #atom - helper1 vector
    v2 = normalize(helper1 - atom)
    #atom - helper2 vector
    v3 = normalize(helper2 - atom)
    #####case(3) !CH2####
    #Perpendicular to the helpers - atom plane
    v4 = normalize(np.cross(v3, v2))
    #Rotational vector
    rot_vec = normalize(v2 - v3)
    #Vector to be rotated by theta/2, perpendicular to rot_vec and v4
    vec_to_rotate = normalize(np.cross(v4, rot_vec))
    norm_vec_H1 = apply_rotation(vec_to_rotate, rot_vec, -1.911/2)
    hcoor_H1 = 1 * norm_vec_H1 + atom
    norm_vec_H2 = apply_rotation(vec_to_rotate, rot_vec, 1.911/2)
    hcoor_H2 = 1 * norm_vec_H2 + atom
    return (hcoor_H1, hcoor_H2)
    
    

#Traduction case(3) fortran subroutine add_hydrogen
##Tests##
#N4, C5, O11, C13
helper1 = np.array([[43.89, 74.55, 21.67], [44.38, 75.92, 21.85], [46.58  ,78.14   ,25.37], [45.10 , 79.68 ,  26.62]])
#C5, C6, C12, C32
atom = np.array([[44.38, 75.92, 21.85],[45.16 , 76.27 , 23.12], [45.77 , 79.32 , 25.29], [44.82 , 78.27,  27.15]])
#C6, O7, C13, O33
helper2 = np.array([[45.16, 76.27, 23.12], [45.59  , 77.63,  23.07], [45.10 , 79.68 ,  26.62], [43.94 , 78.22 , 28.27]])

#Tests of the news functions on simple cases 
index = 1
for i in range(len(atom)):
    (coor_H1, coor_H2) = get_SP2_H(atom[i], helper1[i], helper2[i])
    write_PDB(index, "C", atom[i])
    index += 1 
    write_PDB(index, "H", coor_H1)
    index += 1 
    write_PDB(index, "H", coor_H2)
    




#####case(4) !CH3e
v1 = C5
#ref vectors
#N4-C5 vector
v2 = normalize(N4 - v1)
#C6-C5 vector
v3 = normalize(C6 - v1)

u = normalize(np.cross(v3, v2))

quat_N4_C5 = v2q(u, -1.911/2)

#generate the rotational matrix 
rot_mat_quat_N4_C5 = rotational_matrix(quat_N4_C5)

#Use the rotational matrix on the C5-C6 vector
vec_H51 = np.dot(rot_mat_quat_N4_C5, v2)

norm_vec_H51 = normalize(vec_H51)
hcoor = 0.1 * norm_vec_H51 + C5



